### PR TITLE
modules: samples: Add wasm-micro-runtime as module and add samples

### PR DESCRIPTION
**Pull request by wenyongh**
Add wasm-micro-runtime as a Zephyr module and add two samples:
a hello world sample to illustrate how to run WASM application with
wasm-micro-runtime (WAMR) runtime engine, and an app manager
sample to illustrate how to start WAMR app manager in device
side which runs on Zephyr OS and then install/uninstall/query
WASM apps from host side.

Signed-off-by: Wenyong Huang <wenyong.huang@intel.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
@wenyongh thanks! This seems to address the integration of WASM with Zephyr at the sample level. We should instead try to do something a bit more generic, so that each sample doesn't need to copy over all the boilerplate. What are your thoughts on that?

**Comment by carlescufi** [comment]
@xwang98 could you take a look and compare to your approach?

**Comment by wenyongh** [comment]
> @wenyongh thanks! This seems to address the integration of WASM with Zephyr at the sample level. We should instead try to do something a bit more generic, so that each sample doesn't need to copy over all the boilerplate. What are your thoughts on that?

Yes, it is really not generic and not convenient to developer, how about adding WAMR as a module to Zephyr? For example, add folder zephyr/modules/wasm-micro-runtime and file Kconfig/CMakeLists.txt/wamr_init.c inside it, and define the related macros to configure WAMR features, e.g., CONFIG_WAMR_INTERP, CONFIG_WAMR_AOT? By default the runtime environment is initialized in wamr_init.c, and developer can simply load and run the wasm application. Thanks.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  # Enable libc builtin support by default
+  set (WAMR_BUILD_LIBC_BUILTIN 1)
+endif ()
+
+if (NOT DEFINED WAMR_BUILD_LIBC_WASI)
+  # Disable libc wasi support by default
+  set (WAMR_BUILD_LIBC_WASI 0)
+endif ()
+
+set (WAMR_ROOT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/../../../../modules/lib/wasm-micro-runtime)
```
Comment (tejlmand) [inline samples/wasm_micro_runtime/hello_world/CMakeLists.txt:43]: If the wasm-micro-runtime is made into a Zephyr library, that is having a: `zephyr/module.yml` file with this content:
```
name: wasm-micro-runtime

build:
  cmake-ext: True
  kconfig-ext: True
```
then we can use `ZEPHYR_WASM_MICRO_RUNTIME_MODULE_DIR` instead.

**Review by tejlmand — CHANGES_REQUESTED** [review]
@wenyongh thanks for this contribution. A good start, but needs some additional effort.

I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt (hunk 1 above)
Comment (tejlmand) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt:43]: See also: 
https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/zephyr/module.yml
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/loramac-node/CMakeLists.txt

for inspiration.

**Comment by wenyongh** [comment]
> @wenyongh thanks for this contribution. A good start, but needs some additional effort.
> 
> I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

@tejlmand Thanks a lot for your suggestions! We will modify the patch according your comments and try to make it more generic.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2020 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.8.2)
+
+if (NOT BOARD)
+  set (BOARD qemu_x86_nommu)
+endif ()
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt]: Done, removed.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt]: Done.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+enable_language (ASM)
+
+set (WAMR_BUILD_PLATFORM "zephyr")
+
+# Build as X86_32 by default, change to "AARCH64[sub]", "ARM[sub]", "THUMB[sub]", "MIPS" or "XTENSA"
+# if we want to support arm, thumb, mips or xtensa
+if (NOT DEFINED WAMR_BUILD_TARGET)
+  set (WAMR_BUILD_TARGET "X86_32")
+endif ()
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt]: Done, we added CONFIG_WAMR_BUILD_TARGET to allow user to specify the build target, and it isn't set, the build scripts will configure it automatically according to the CPU architecture info. Normally developer doesn't need to set it.

File: samples/wasm_micro_runtime/hello_world/README.rst [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+To build and run the wasm app in other targets:
+
+(1) stm32
+
+.. code-block:: console
+
+    west build -b nucleo_f767zi . -p always -- \
+               -DCONF_FILE=target_configs/prj_nucleo767zi.conf \
+               -DWAMR_BUILD_TARGET=THUMBV7
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/README.rst]: Done, -DWAMR_BUILD_TARGET=xxx was removed.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+if (NOT DEFINED WAMR_BUILD_LIBC_BUILTIN)
+  # Enable libc builtin support by default
+  set (WAMR_BUILD_LIBC_BUILTIN 1)
+endif ()
+
+if (NOT DEFINED WAMR_BUILD_LIBC_WASI)
+  # Disable libc wasi support by default
+  set (WAMR_BUILD_LIBC_WASI 0)
+endif ()
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt]: Done, the related configs were moved to wasm-micro-runtime module's Kconfig.

File: samples/wasm_micro_runtime/hello_world/CMakeLists.txt (hunk 1 above)
Comment (wenyongh) [reply samples/wasm_micro_runtime/hello_world/CMakeLists.txt]: Done.

**Comment by wenyongh** [comment]
> @wenyongh thanks for this contribution. A good start, but needs some additional effort.
> 
> I agree with @carlescufi that this should be more generic, and for a start I have placed a few comments for you to take a look at.

@tejlmand @carlescufi Many thanks for your suggestions, we have updated the PR to make WAMR as a zephyr module and updated the related samples' build scripts and documents, could you please help review the patch again? And could you please fork [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) into [zephyrproject-rtos](https://github.com/zephyrproject-rtos) so that we can update zephyr/west.yml? Thanks.

**Comment by wenyongh** [comment]
> @wenyongh one thing I am missing here, where is the actual code being pulled from? We should likely be adding an entry to `west.yml`? Or is one supposed to download the runtime separately?

Yes, it supposes that the WAMR source code is forked into https://github.com/zephyrproject-rtos and west will clone it under zephyrproject/modules/lib. I have updated zephyr/west.yml to add wasm-micro-runtime into project list. The patch should work after [WAMR](https://github.com/bytecodealliance/wasm-micro-runtime) project is cloned.

**Review by tejlmand — CHANGES_REQUESTED** [review]
First round of comments

**Comment by ivan2angelov** [comment]
I am very interested in studying WASMR as part of Zephyr. This PR looks almost complete. Are there any plans to merge it?
bot closed it due to lack of activity, @wenyongh shall this PR be reopen?

**Comment by PureEngineering** [comment]
I don't understand why this wasn't merged in? can someone revisit this.

**Comment by xwang98** [comment]
This is the Zephyr team's call for accepting this PR or not.  @ivan2angelov @PureEngineering

**Comment by ivan2angelov** [comment]
@xwang98 do you have some contact to Zephyr's team? Is it possible to raise an issue to adding WAMR to Zephyr?

**Comment by soburi** [comment]
@wenyongh @carlescufi @tejlmand 

I'm interested in this PR.
At a glance, all pointing outs are solved.
And I verified this PR can work on the newest version of Zephyr with few modifications.

So, what reason to reject this PR at that time?
I will help to solve if there are problems.

**Pull request closed without merging**